src/ray/object_manager/pull_manager.cc

```diff
@@ -122,57 +156,87 @@ void PullManager::UpdatePullsBasedOnAvailableMemory(size_t num_bytes_available)
     RAY_LOG(DEBUG) << "Updating pulls based on available memory: " << num_bytes_available;
   }
   num_bytes_available_ = num_bytes_available;
-
-  // While there is available capacity, activate the next pull request.
   std::vector<ObjectID> objects_to_pull;
-  while (num_bytes_being_pulled_ < num_bytes_available_) {
-    // Get the next pull request in the queue.
-    const auto last_request_it = pull_request_bundles_.find(highest_req_id_being_pulled_);
-    auto next_request_it = last_request_it;
-    if (next_request_it == pull_request_bundles_.end()) {
-      // No requests are active. Get the first request in the queue.
-      next_request_it = pull_request_bundles_.begin();
-    } else {
-      next_request_it++;
+  std::unordered_set<ObjectID> object_ids_to_cancel;
+  // If there are any worker requests queued, try to activate them. Since we
+  // prioritize worker requests over task requests, task requests will be
+  // canceled as necessary to make space. We may exit this block over capacity
+  // if we run out of task requests to cancel, but this will be remedied later
+  // by canceling worker requests.
+  bool worker_requests_remaining = !worker_request_bundles_.empty();
+  while (worker_requests_remaining) {
+    worker_requests_remaining = ActivateNextPullBundleRequest(
+        worker_request_bundles_, &highest_worker_req_id_being_pulled_, &objects_to_pull);
+    while (num_bytes_being_pulled_ > num_bytes_available_) {
+      // Activating the last worker request put us over the available memory.
+      // Try to cancel task requests to make space.
+      if (highest_task_req_id_being_pulled_ == 0) {
```

Comment:
```suggestion
      if (highest_task_req_id_being_pulled_ <= 0) {
```

Consider this defensive change

Reply:
The req_ids are all unsigned.